feat(Caching): Cache across Hierarchy Keyring (#279)
awslabs/aws-encryption-sdk-specification: changes/2024-TBD_cache-across-hierarchy-keyrings/change.md

Reviewer: I am not reviewing the change doc today.

- Hierarchy Keyring: keyStore#BranchKeyMaterials
- CachingCMM: EncryptionMaterials and DecryptionMaterials
- DB-ESDK: keyStore#BeaconKeyMaterials

These materials have certain cache identifiers for accessing
them until a TTL. As of [today](#today), the cache identifiers
are set up such that:

- Only one Hierarchy Keyring can use one cache,
  i.e., currently, we do NOT support a shared cache
  between multiple Hierarchy Keyrings.
- Multiple CachingCMMs can use a single shared cache
  (this is for native languages, there is currently no
  CachingCMM implemented in dafny)
- We do NOT support a shared cache between multiple
  DB-ESDK clients.

We will focus on the Hierarchy Keyring unless mentioned
otherwise. Our aim is to support a shared cache between
multiple Hierarchy Keyrings
as highlighted in the next sections.

## Motivation

The Hierarchy Keyring,
and it's component the (Branch) Keystore,
allow MPL Consumers to reduce their KMS Call volume
by persisting KMS protected cryptographic materials into
an available medium
(currently, only a DynamoDB table is available as persistence medium).

We call these cryptographic materials Branch Keys.

> However, an instance of the Hierarchy Keyring

Reviewer 2: does this mean, multiple keyrings may have the same KMS relationship?

Author: Yes, you can create multiple keyrings with the same KMS relationships.

Reviewer 2: so then wouldn't multiple keyrings (sharing the same keystore) create cache IDs with collision?

when using the Hierarchy Keyring,
we need to allow caching across KeyStores/KMS Clients/KMS Keys.

To facilitate Caching across KeyStores/KMS Clients/KMS Keys,
we MUST break the Cryptographic Materials Cache (CMC)
out of the Hierarchy Keyring.

By allowing MPL Consumers to provide an already initialized CMC
to the Hierarchy Keyring at construction,
the CMC MAY cache Branch Keys protected by different
KMS Relationships.

This simplifies Multiple KMS Relationship MPL Consumers,
as they do not need to stand up a LRU Cache of Hierarchy Keyrings.

Instead, they may maintain one CMC.
They still create a Hierarchy Keyring instance per KMS Relationship,
and they MUST use the correct Keyring to retrieve material
from the Cache.

But they only need to maintain the common cache.

In the future, the CachingCMM will be introduced to
Crypto Tool's Dafny products.
The CachingCMM and the Hierarchy Keyring both consume CMCs.
Thus, it will be possible to provide a CMC to both a
Hierarchy Keyring & a CachingCMM.
This cache identifier design MUST consider
how the cache entries of multiple Hierarchy Keyrings
and CachingCMMs will be appropriately separated.

## Requirements

> - We must allow caching across KeyStores/KMS Clients/KMS Keys

Reviewer: nit: We have switched from KeyStore/keystore to Key Store. 
Two words.

Author: Okay, thanks!

  or request headers

The Local Cryptographic Material Cache of
the Hierarchy Keyring instance is then only
populated with Branch Keys that correspond with
that KMS relationship.

Which is appropriate,
as it is clear under what KMS relationship
a Branch Key is accessed.

However,
the Hierarchy Keyring,
and it's Key Store,
have a runtime cost,
exerting memory pressure
and, without manual optimization,
requiring at least 2 TLS handshakes
when first serving a request
(TLS to KMS & TLS to DDB).

Additionally,
the local Cryptographic Materials Cache
exerts some runtime cost,
particularly in a multi-threaded environment,
when a background worker thread MAY be refreshing
or pruning entries of the cache.

For MPL Consumers that MUST work with Branch Keys
under different KMS Relationships,
this runtime cost adds up.

These MPL Consumers MAY end up establishing
a LRU Cache of Hierarchy Keyrings.
Which, while workable, is sub-optimal,
and clearly makes the Hierarchy Keyring,
in these conditions,
"Hard to Use".

The objective, with these changes,
is to make the Hierarchy Keyring
"Easy to Use" in a multiple KMS Relationship
environment.

## Summary

> Relevant [sim](https://sim.amazon.com/issues/CrypTool-5311).

Author: Great point. Thanks!

  for multiple Hierarchy Keyrings and CachingCMMs.
  For this, we need to strategically update the cache identifiers
  for all materials ([background](#background)) in the CMC.
  The [Issues and Alternatives section](#issues-and-alternatives)
  discusses this.

## Out of Scope

- Supporting a shared Cache in the DB-ESDK to fetch
  BeaconKeyMaterials across multiple Searchable
  Encryption Configurations, or across multiple KMS Relationships.

## Issues and Alternatives:

Before looking at the issues, please take a quick look at
[Hierarchy Keyring Cache Identifier formula](#hierarchy-keyring-cache-identifier-formula),
[CachingCMM Cache Identifier formula](#cachingcmm-cache-identifier-formula), and
[DB-ESDK Searchable Encryption Cache Identifier formula](#db-esdk-searchable-encryption-cache-identifier-formula) for more context.

Note that there are two hashings taking place.
For example in the Hierarchy Keyring Cache Identifier formula,
one hash is for the branch-key-id and the other is hashing
the entire appended string
(a part of which is the hash of the branch-key-id)
to get the final cache identifier.
We'll call these internal and external hashes.

> #### Issue 1: How to update the cache identifier for the Hierarchy Keyring to allow a shared cache?

Reviewer: nit/formatting: you went from two `#` to four `#`, instead of 3 `#`.
I like neat hierarchies, 
but this 100% an author's choice.

Author: I had another heading there that I removed in the subsequent draft. Thanks for pointing this out.

Note that according to the current
[Hierarchy Keyring Cache Identifier formula](#hierarchy-keyring-cache-identifier-formula),
if two KMS Relationships Hierarchy Keyrings are
called with two different Key Stores but the same branch-key-id,
there will be a collision (the branch-key-version will also
have to co-incide in case of Decryption Branch Key Materials).

> ##### Option 1 (Recommended): Append the keystore-id to the [Hierarchy Keyring cache identifier formula](#hierarchy-keyring-cache-identifier-formula).

Reviewer 2: what if keystores are shared between keyrings?

Author: Moving to using partitionId instead of keyStoreId

> By appending the keystore-id to the

Author: Great point. Thanks

Hierarchy Keyring cache identifier,
we can ensure that there are no collisions between two identifiers

> from different Key Stores.

Author: Okay, sounds good. Thanks.

Adding a slightly modified version of this to the doc based on our slack conversation
'By adding Key Store ID, we let the customer control collisions.
I.e: they choose if different KMS Relationships CAN hit the same identifier or not.
They can also choose if different logical key stores or physical key stores can collide or not.
But by default, because each Key Store object will (likely) have a unique UUID,
there will be no collision.'

More context on this can be found in the
[Appendix section A1](#a1-relating-cache-entries-branch-keys-to-kms-relationships).

##### Option 2: Append "unique numbers" to the cache identifiers of different keyrings

This is not practical for a multi-threaded system.

> Also, we cannot keep track of these "unique numbers".

Reviewer 2: doesn't this option differ from option 1 because the keystore-id is "unique numbers" that we CAN track?

> Also, we cannot keep track of these "unique numbers".

Reviewer 2: doesn't this option differ from option 1 because the keystore-id is "unique numbers" that we CAN track?

> Also, we cannot keep track of these "unique numbers".

Reviewer: As I have read this:
- Option 1 is use the Key Store ID appended to the rest of the cache identifier.
- Option 2 is use a unique number appended to the rest of the cache identifier.

But, if the default value for Key Store ID is a UUID generated at local Key Store construction time,
how is this UUID practically different than a unique number?

> Also, we cannot keep track of these "unique numbers".

Author: Good point. Thanks.

> Also, we cannot keep track of these "unique numbers".

Author: Both of you are correct. Key Store IDs default to unique UUIDs that we CAN track. So we're gonna use Key Store ID (which default to a randomly generated UUID). In addition to that, a customer can set the Key Store ID to allow cache sharing between different KMS configurations. But by default, the Key Store IDs are unique, which is perfect for our use case.

has branch-key-id appended in the cache identifier formula and the
[Branch Key Materials (Encryption) cache identifier](https://github.com/awslabs/aws-encryption-sdk-specification/tree/ffb2b0cc6a956b2cec3a33be3c3672605b6907fb/framework/aws-kms/aws-kms-hierarchical-keyring.md#encryption-materials)
has branch-key-digest=SHA512(branch-key-id). Here we're talking
about the internal hash.

##### Option 1 (Recommended): Update the Decryption cache identifier to include branch-key-digest=SHA512(branch-key-id) instead of the branch-key-id

This will make both the Hierarchy Keyring Encryption and
Decryption cache identifiers more coherent.
The cache identifier input would be deterministic
irrespective of the size of branch-key-id that the
customer provides.
Additionally, if we're updating branch-key-id to
branch-key-digest, we can remove the branch-key-id-length
parameter in the formula and make it cleaner.
This has CryptoBR's approval
[[ticket]](https://issues.amazon.com/issues/P148158466).

##### Option 2: Update the Encryption cache identifier to include branch-key-id instead of the branch-key-digest=SHA512(branch-key-id)

> Not hashing the branch keys will mean that the input

Reviewer 2: not hashing the branch key **ids**

Author: Good catch. Thanks.

to the final hash function will be of variable length,
which is hard to reason about.

##### Option 3: Don't change anything.

This would mean that the formula will remain incoherent.

#### Issue 3: Which hashing algorithm (internal and external) should be used in the Hierarchy Keyring cache identifier and the CachingCMM?

Note that according to the
[cache identifier formulae](#cache-identifier-formulae),
as of [today](#today), the length of cache identifier for
BranchKeyMaterials in Hierarchy Keyring is 32 bytes for
both Encrypt and Decrypt and the length of cache identifier
for EncryptionMaterials and DecryptionMaterials in the
CachingCMM is 64 bytes.
Both Hierarchy Keyring and CachingCMM identifiers currently
use SHA512, but the Hierarchy Keyring cache identifier is
truncated to 32 bytes.

##### Option 1 (Recommended): Using SHA512 for both

SHA512 has more number of bits to represent the output
of the cache identifier strings. More information about
why this is true based on the birthday bound can be found
[here](#birthday-bound-for-sha512-and-sha384)

> This is pending [CryptoBR guidance](https://t.corp.amazon.com/P148158466/communication).

Author: Yes, I'll do that. Thanks.

##### Option 2: Change the hashing algorithm to SHA384 in both

SHA512 is the better choice as mentioned above.
More information about this can be found
[here in the appendix](#a3-information-about-sha384-and-length-extension-attacks),
where I discuss length extension attacks and why

> SHA512 is still the better choice.

Author: Yes, that's a good point. I've updated this now. Thanks.

a Hierarchy Keyring and a CachingCMM idenitifier will
be distinct if we add the UTF8 encodings of prefixes
'hierarchy_keyring' (17 bytes) and 'caching_cmm' (11 bytes)
to each of them respectively.

### Distinct sizes of the identifiers

As an additional safety measure, we look closely
at the formulae for cache identifiers in a Hierarchy Keyring
and a CachingCMM encrypt and decrypt (note that these will
actually be a total of 5, since the
[cachingCMM encrypt cache identifier](https://github.com/awslabs/aws-encryption-sdk-specification/tree/ffb2b0cc6a956b2cec3a33be3c3672605b6907fb/framework/caching-cmm.md#appendix-a-cache-entry-identifier-formulas)
has two different formulae based on if we
specify the algorithm suite or not).

In the
[Hierarchy Keyring Cache Identifier formula](#hierarchy-keyring-cache-identifier-formula)
Encrypt, as discussed before, we will remove the len_branch_key.
For Decrypt, we will remove the len_branch_key
and use the branch-key-digest in place of the branch-key-id.
The size of the inputs to the external hash for both these
cases then becomes:

- Encrypt: 17 (prefix) + 64 + 1 + 6 = 88 bytes

> - Decrypt: 17 + 64 + 1 + 36 = 118 bytes

Reviewer: Issue: You need to specify which of these bytes are constant and which are variable.
For example, the prefixes are constant.
But the branch-key-digest is variable.

Author: Great idea. Thanks.

Similarly, after making the proposed updates in the
[CachingCMM Cache Identifier formula](#cachingcmm-cache-identifier-formula),
the size will be:

- Encrypt without Algorithm Suite specified: 11 (prefix) + 64 + 1 + 64 = 140 bytes
- Encrypt with Algorithm Suite specified: 11 + 64 + 1 + 2 + 64 = 142 bytes
- Decrypt: 11 + 64 + 2 + 64\*x + 64 + 64 = 205 + 64x bytes ; where x is the number of EDKs in the decrypt string

In addition to the explanations in the previous
[subsection](#derivation-from-assumptions),
all these input strings are also unique
because they have different lengths. We just need
to choose a good hashing algorithm which gives us a
very large output space to project these strings to
minimize collisions, and SHA512 will do the job.

## Cache Identifier Formulae

### Hierarchy Keyring Cache Identifier formula

> - [Encrypt](https://github.com/aws/aws-cryptographic-material-providers-library/blob/c07a51fc29ff70411f7573bca96d2a091db8c1ed/AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/src/Keyrings/AwsKms/AwsKmsHierarchicalKeyring.dfy#L384) - SHA512(len_branch_key + SHA512(branch_key_id_utf8) + [0x00] + activeUtf8)[0..32] [[spec]](https://github.com/awslabs/aws-encryption-sdk-specification/tree/ffb2b0cc6a956b2cec3a33be3c3672605b6907fb/framework/aws-kms/aws-kms-hierarchical-keyring.md#encryption-materials)

Reviewer 2: len_branch_key is misleading. I believe it should refer to the length of the string branch-key-id, which is Uint32, as bytes

Author: That's a good point. However, we're going to remove the length of the branch key ID from the cache identifier because it doesn't add anything.

  - The SHA512(branch_key_id_utf8) is called the branch-key-digest

> - [Decrypt](https://github.com/aws/aws-cryptographic-material-providers-library/blob/c07a51fc29ff70411f7573bca96d2a091db8c1ed/AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/src/Keyrings/AwsKms/AwsKmsHierarchicalKeyring.dfy#L725) - SHA512(len_branch_key + branch_key_id_utf8 + [0x00 as uint8] + UTF8.EncodeAscii(branchKeyVersion))[0..32] [[spec]](https://github.com/awslabs/aws-encryption-sdk-specification/tree/ffb2b0cc6a956b2cec3a33be3c3672605b6907fb/framework/aws-kms/aws-kms-hierarchical-keyring.md#decryption-materials)

Reviewer 2: [same](https://github.com/awslabs/aws-encryption-sdk-specification/pull/279/files#r1720116498)

## Birthday Bound for SHA512 and SHA384

A [birthday attack](https://en.wikipedia.org/wiki/Birthday_attack)
is a bruteforce collision attack that exploits the
mathematics behind the birthday problem in probability
theory. This attack can be used to abuse communication
between two or more parties. The attack depends on the
higher likelihood of collisions found between random
attack attempts and a fixed degree of
permutations (pigeonholes). With a birthday attack,
if we have a total of n bits output and therefore
$2^n$ different outputs, it is possible to find a
collision of a hash function with 50% chance by
calculating approximately $sqrt(2^n) = 2^{n/2}$ hashes.

To find a random collision in a SHA512 hash with
probability 0.5, we would need to calculate
approximately $2^256$ hashes. Similarly for SHA384,
we would need to calculate $2^192$ hashes.
Therefore, SHA512 is more collision resistant.

## Appendix

## A1: Relating Cache Entries (Branch Keys) to KMS Relationships

For some MPL Consumers,
Cache Entries SHOULD only be served IF
the request is being served under the same KMS Relationship,
which is the most secure condition,
as it prevents a Confused Deputy scenario

> (detailed in **Confused Deputy 1**).

Reviewer: Issue/Suggestion: Remove.
I never wrote this; it was on my TODO list.
Alternatively, if you want to write, you can. 
But I do not think it is needed today.
```suggestion
(detailed in **Confused Deputy 1**).
```

Author: I'll remove this as there are more pertinent things to focus on.

- Minor update to the spec: The branch-key-version (which is the UUID) in the [Hierarchy Keyring Branch Key Materials (Decryption) cache identifier](https://github.com/awslabs/aws-encryption-sdk-specification/tree/ffb2b0cc6a956b2cec3a33be3c3672605b6907fb/framework/aws-kms/aws-kms-hierarchical-keyring.md#decryption-materials) is currently mentioned to be of size 36 bytes. However, the UUID is actually only 16 bytes. We get 36 bytes when we convert the 16 bytes in UUID to a UTF8 encoding. Therefore, the spec should mention that the branch-key-version has size 16 bytes [[ref: Hierarchy Keyring Cache Identifier formula]](#hierarchy-keyring-cache-identifier-formula).

## A3: Information about SHA384 and Length Extension Attacks

SHA384 computation speed is the same as that of SHA512 [[ref](https://news.ycombinator.com/item?id=27960348#:~:text=For%2064%2Dbit%20CPUs%20without,the%20algorithm%20is%20the%20same)].

However, SHA384 is not prone to
[Length Extension attacks](https://en.wikipedia.org/wiki/Length_extension_attack),
unlike SHA512, which is a plus for SHA384.
But, cache identifiers don't fall into the category

> of length extension attacks, which come into the picture

Reviewer 3: The reason cache identifiers may have issues is that if you can create a collision, then you would have 2 different branch key versions that are cached in the "same place".

Consider, you can select any branch key name you want.
So you select one such that the digest of your branch key will collide with mine.
You then make sure that you make a write request before me.
Now all my data will be encrypted with your branch key.

This is because the cache will be hydrated with you branch key,
and when the system looks to see if my branch key is cached
it will find an entry.

This can be mitigated in several ways.
But it is something to think about.

Author: There are two kinds of threats I can see here.

First, if both users use different key stores. If that is the case, the key store IDs will be different (unless the customer explicitly wants / sets them to be the same. 

Second, if they use the same key store, then, if I remember correctly, there can't be two branch key ids with the same name. The digest is a SHA512, so it is safe to assume that if the branch key ids are different, the digests will be different, given the large output space (2^512).

Reviewer: Nit: One threat, two exploits/mitigations, depending on how you look at it.

But I think Ryan is on to something;
we need to describe the threat in this doc.

Author: This should be fixed with the updated with the most recent cache identifiers: `partitionId + logicalKeyStoreName + branchKeyId`

This serves as a reference of all specification documents that this change affects.

| Specification                                                                           |
| --------------------------------------------------------------------------------------- |
| [Key Store](../../framework/branch-key-store.md)                                        |
| [AWS KMS Hierarchical Keyring](../../framework/aws-kms/aws-kms-hierarchical-keyring.md) |

## Affected Implementations

| Language | Version Introduced | Version Removed | Implementation                                                                                                                                                                                                                                                                                      |
| -------- | ------------------ | --------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | Dafny    | 1.4.0              | n/a             | [AwsKmsHierarchicalKeyring.dfy](https://github.com/aws/aws-cryptographic-material-providers-library/blob/main/AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/src/Keyrings/AwsKms/AwsKmsHierarchicalKeyring.dfy)                                                          |

Reviewer 4: Dafny 1.4.0 ?

Author: Discussed - This is the MPL version

requiring at least 2 TLS handshakes
when first serving a request
(TLS to KMS & TLS to DDB).

Additionally,
the local Cryptographic Materials Cache
exerts some runtime cost,
particularly in a multi-threaded environment,
when a background worker thread MAY be refreshing
or pruning entries of the cache.

For MPL Consumers that MUST work with Branch Keys
under different KMS Relationships,
this runtime cost adds up.

These MPL Consumers MAY end up establishing
a LRU Cache of Hierarchy Keyrings.
Which, while workable, is sub-optimal,
and clearly makes the Hierarchy Keyring,
in these conditions,
"Hard to Use".

The objective, with these changes,
is to make the Hierarchy Keyring
"Easy to Use" in a multiple KMS Relationship
environment.

## Summary

<!-- TODO: Remove SIM before merging to mainline -->

Relevant [sim](https://sim.amazon.com/issues/CrypTool-5311).

To improve the MPL Consumer data key caching experience,
when using the Hierarchy Keyring,
we need to allow caching across Key Stores/KMS Clients/KMS Keys.

To facilitate Caching across Key Stores/KMS Clients/KMS Keys,
we MUST break the Cryptographic Materials Cache (CMC)
out of the Hierarchy Keyring.

> By allowing MPL Consumers to provide an already initialized CMC
> to the Hierarchy Keyring at construction,
> the CMC MAY cache Branch Keys protected by different
> KMS Relationships.

Reviewer 5: Adding this cache MUST be an optional parameter, right? Otherwise current customers of the keyring will get compile errors saying, "must provide cache in initialization" or something like that. 
```suggestion
By allowing MPL Consumers to optionally provide an initialized CMC
to the Hierarchy Keyring at construction,
the CMC MAY cache Branch Keys protected by different
KMS Relationships.
```

Author: Yes, you're right. It is an optional parameter that the customer can use.

This simplifies Multiple KMS Relationship MPL Consumers,
as they do not need to stand up a LRU Cache of Hierarchy Keyrings.

Instead, they may maintain one CMC.
They still create a Hierarchy Keyring instance per KMS Relationship,
and they MUST use the correct Keyring to retrieve material

> from the Cache.

Reviewer 4: Does this mean we are exposing an interface to accept a Cache while the internal working remains (mostly) the same? Or are we changing how materials are cached as well (excluding the identifier change).

Author: First. The cache interface does NOT change. We are only changing the way we take a shared cache as input.

But they only need to maintain the common cache.

In the future, the CachingCMM will be introduced to
Crypto Tool's Dafny products.
The CachingCMM and the Hierarchy Keyring both consume CMCs.
Thus, it will be possible to provide a CMC to both a
Hierarchy Keyring & a CachingCMM.
This cache identifier design MUST consider
how the cache entries of multiple Hierarchy Keyrings

> and CachingCMMs will be appropriately separated.

Reviewer 5: If this spec change is going to propose a change to caching cmm cache id formulas we should say that this cahnge also impacts the caching cmm up above or say that it is out of scope in the below section.

Author: We are updating the spec to update the cachingCMM cache identifiers.

Currently, we're only focussing on making the changes to the MPL. Eventually, everything is going to be moved to Dafny anyway.

If a case can be made for other native libraries, we can incorporate that.

Author: Update: Native library cache identifiers can NEVER collide with the new proposed cache identifiers.

  discusses this.

## Out of Scope

- Supporting a shared Cache in the DB-ESDK to fetch
  BeaconKeyMaterials across multiple Searchable
  Encryption Configurations, or across multiple KMS Relationships.

## Issues and Alternatives:

Before looking at the issues, please take a quick look at
[Hierarchy Keyring Cache Identifier formula](#hierarchy-keyring-cache-identifier-formula),
[CachingCMM Cache Identifier formula](#cachingcmm-cache-identifier-formula), and
[DB-ESDK Searchable Encryption Cache Identifier formula](#db-esdk-searchable-encryption-cache-identifier-formula) for more context.

Note that there are two hashings taking place.
For example in the Hierarchy Keyring Cache Identifier formula,
one hash is for the branch-key-id and the other is hashing
the entire appended string
(a part of which is the hash of the branch-key-id)
to get the final cache identifier.
We'll call these internal and external hashes.

### Issue 1: How to update the cache identifier for the Hierarchy Keyring to allow a shared cache?

Note that according to the current
[Hierarchy Keyring Cache Identifier formula](#hierarchy-keyring-cache-identifier-formula),
if two KMS Relationships Hierarchy Keyrings are

> called with two different Key Stores but the same branch-key-id,

Reviewer 5: we may want an example here. As I was reading; I said to myself, "ok, so? but then I was like right the KMS relationships above talk about clients with different regions and creds and so on"

Author: Resolving this. Removed the design doc and created a background / change doc. I don't think we need this examples there.

there will be a collision (the branch-key-version will also
have to co-incide in case of Decryption Branch Key Materials).

#### Option 1 (Recommended): Append the keystore-id to the [Hierarchy Keyring cache identifier formula](#hierarchy-keyring-cache-identifier-formula).

> By appending the [`keystore-id`](../../framework/branch-key-store.md#keystore-id) to the

Reviewer 5: hmm what if I create my keystore with a custom keystore-id?
Since it could be supplied this would or wouldn't work?

Author: Update: Clarify between accidental collisions and explicit cache hits.

Author: Should be clarified now with the latest updates.

Hierarchy Keyring cache identifier,
we can ensure that there are no collisions between two identifiers
from different local Key Stores.

> Key Store IDs are UUIDs that are unlikely to collide by default.

Reviewer: nit: 
```suggestion
By default, Key Store IDs are UUIDs generated at object initialization; 
they are unlikely to collide unless they are manually set.
```

> Key Store IDs are UUIDs that are unlikely to collide by default.

Reviewer 6: User can also manually send the Key Store ID according to current spec. Are we defaulting it to UUID?

Reviewer 6: Tony's suggestion answered my question.

> Key Store IDs are UUIDs that are unlikely to collide by default.

Reviewer 4: I thought they were the name of the DDB Table. How this UUID is generated?

Author: We moved to using partition Id instead of Key Store Id. But both these are modeled in the same way. If the partition Id is provided by the customer, we set the partition Id to utf8 encoding of the input. Otherwise, we set it to a randomly generated 16 byte UUID.

By adding Key Store ID to the Hierarchy Keyring cache identifier,
the customer can control cache access.
For instance, they can choose if different
KMS Relationships CAN hit the same identifier or not.
They can also choose if different logical key stores
or physical key stores can collide or not.
But by default, because each Key Store object will have a unique UUID,
there will be no collision.
More context on this can be found in the
[Appendix section A1](#a1-relating-cache-entries-branch-keys-to-kms-relationships).

> #### Option 3: Don't change anything.

Reviewer 4: Where's Option 2?

Author: Fixed

> #### Option 3: Don't change anything.

Reviewer 5: ```suggestion
#### Option 2: Don't change anything.
```

from multiple Hierarchy Keyrings.

### Issue 2: Should we update the cache identifier for the Hierarchy Keyring BranchKeyMaterials to include branch-key-digest instead of branch-key-id?

Note that according to the current
[Hierarchy Keyring Cache Identifier formula](#hierarchy-keyring-cache-identifier-formula),
the [Hierarchy Keyring Branch Key Materials (Decryption) cache identifier](https://github.com/awslabs/aws-encryption-sdk-specification/tree/ffb2b0cc6a956b2cec3a33be3c3672605b6907fb/framework/aws-kms/aws-kms-hierarchical-keyring.md#decryption-materials)
has branch-key-id appended in the cache identifier formula and the
[Branch Key Materials (Encryption) cache identifier](https://github.com/awslabs/aws-encryption-sdk-specification/tree/ffb2b0cc6a956b2cec3a33be3c3672605b6907fb/framework/aws-kms/aws-kms-hierarchical-keyring.md#encryption-materials)
has branch-key-digest=SHA512(branch-key-id). Here we're talking
about the internal hash.

#### Option 1 (Recommended): Update the Decryption cache identifier to include branch-key-digest=SHA512(branch-key-id) instead of the branch-key-id

This will make both the Hierarchy Keyring Encryption and

> Decryption cache identifiers more coherent.

Reviewer 7: nit: Do you mean consistent? I think the formulas are coherent but not consistent.

CachingCMM is 64 bytes.
Both Hierarchy Keyring and CachingCMM identifiers currently
use SHA512, but the Hierarchy Keyring cache identifier is
truncated to 32 bytes.

#### Option 1 (Recommended): Using SHA512 for both

SHA-512 has a greater range (512 bits) than
a truncated (to 32 bytes) SHA-512 (256 bits)
or SHA-384 (384 bits). The greater the range,
the less likely there is a collision.
More information about why this is true based
on the birthday bound can be found
[here](#birthday-bound-for-sha512-and-sha384).

<!-- TODO: Remove line before merging to mainline -->

This is pending [AWS Cryptographers' guidance](https://t.corp.amazon.com/P148158466/communication).

#### Option 2: Change the hashing algorithm to SHA384 in both

SHA512 is the better choice as mentioned above.
However, AWS cryptographer's generic recommendation for
Hash Functions is SHA-384.
We believe they make this recommendation based on criteria
other than collision resistance.

<!-- TODO: Once CryptoBR has confirmed SHA-512 >> SHA-384 for
this purpose, revise this section to state "We know" instead of "We believe" -->

More information on length extension attacks
in SHA384 and SHA512 can be found

> [here in the appendix](#a3-information-about-sha384-and-length-extension-attacks).

Reviewer: Nit: This "comes out of the blue".
What is missing, that you and I know, is that the SHA-512 is truncated to mitigate length extension attacks.

Maybe remove this sentence, 
but leave the appendix,
and add to the appendix that the original H-Keyring Formula 
mitigated length extension attacks,
but that mitigation is not needed.

Author: Good point. I'll do that. Thanks.

#### Option 3: Using different hashing algorithms

This would mean different cache identifiers will be
unnecesserily incoherent. Also this deviates from
AWS Cryptographers' guidance.

### Issue 4: Should we add a prefix to all cache identifiers in Hierarchy Keyrings and CachingCMMs?

Here, by prefix we mean, adding the UTF8 encoding
of the string 'hierarchy_keyring' as a
prefix to a hierarchy keyring cache identifier,
and adding encoding of 'caching_cmm' as a prefix to a
CachingCMM cache identifier.

> #### Option 1 (Recommended): Yes

Reviewer 4: I feel this is trivial. If you are not trusting your hash function then there are bigger issues, which is beyond adding a prefix.

Author: Updated the cache identifiers. This shouldn't be an issue now.

Note that after adding Key Store ID / partition ID to a
Hierarchy Keyring's cache identifier, multiple Hierarchy Keyrings
can use a shared cache. Multiple CachingCMMs can use a
shared cache already as mentioned in the [background](#background).
Adding prefixes is an additional safety measure to prevent any
possible collisions between a
Hierarchy Keyring and a CachingCMM.

#### Option 2: No

If we don't add prefixes, even though the chance of an
overlap is minimal, we might get collisions between a
Hierarchy Keyring and a CachingCMM cache identifier.
Adding prefixes adds another layer of distinction between them.

## How can we ensure cache identifiers are collision resistent?

To answer this question, I will use two approaches.

### Derivation from assumptions

> For the HKeyring and CachingCMM, ensuring that all cache

Reviewer 5: nit: I think this is the first time you directly reference HKeyring, which is fine, but when you first mention the Hierarchy Keyring add (HKeyring) afterwards to make it clear we are shortening it.

Two CachingCMMs already allow shared cache.

Per issue
[4](#issue-4-should-we-add-a-prefix-to-all-cache-identifiers-in-hierarchy-keyrings-and-cachingcmms),
a Hierarchy Keyring and a CachingCMM idenitifier will
be distinct if we add the UTF8 encodings of prefixes
'hierarchy_keyring' (17 bytes) and 'caching_cmm' (11 bytes)
to each of them respectively.

### Distinct sizes of the identifiers

As an additional safety measure, we look closely
at the formulae for cache identifiers in a Hierarchy Keyring
and a CachingCMM encrypt and decrypt (note that these will
actually be a total of 5, since the
[cachingCMM encrypt cache identifier](https://github.com/awslabs/aws-encryption-sdk-specification/tree/ffb2b0cc6a956b2cec3a33be3c3672605b6907fb/framework/caching-cmm.md#appendix-a-cache-entry-identifier-formulas)
has two different formulae based on if we
specify the algorithm suite or not).

In the
[Hierarchy Keyring Cache Identifier formula](#hierarchy-keyring-cache-identifier-formula)
Encrypt, as discussed before, we will remove the len_branch_key.
For Decrypt, we will remove the len_branch_key
and use the branch-key-digest in place of the branch-key-id.

> We also add the keystore-id, which is a UUID,

Reviewer: issue: keystore-id defaults to a UUID, but it can be user input... so it can be `n` length.

Author: Ohh yes you're right.

Because we want the cache identifier to be of uniform length, the best way to move forward would be to hash the Key Store ID.

I'll update this in the doc after the meeting.

Reviewer: Maybe? 
Alternatively, we could "invent" a new field for the Key Store object;
key store digest.

The Key Store digest is either:
- UUID, if no Key Store ID is given
- SHA(Key Store ID) if Key Store ID is given

If we can find a SHA that gives 16 bytes, you end up with 16 bytes either way.

intelligently.

For example, imagine we have `N` KMS Relationships,
one of which is `X`,
which are used when retrieving Branch Keys, `n'`,
one of which is `x'`.

A common Cryptographic Materials Cache is established
across all `N` Hierarchical Keyrings used to serve `n'`.

But the Keyrings are routinely garbage collected,
only the common cache remains "in scope" indefinitely.

Whenever `x'` is needed,
a new KMS Client is created that respects relationship `X`,
which is then used to create Key Store Identified as `X`,
and the Key Store is used to create a Hierarchical Keyring
that uses the common Cache.

All entries in this common Cache from that hierarchal Keyring
MUST have a cache identifier that uniquely represents `X`.

When the Hierarchical Keyring for `X` gets garbage collected,
the cache entries MAY remain.

But to retrieve an entry,
the MPL Consumer SHOULD recreate

> a KMS Client is created that respects relationship `X`,
> which is then used to create Key Store Identified as `X`,
> and the Key Store is used to create a Hierarchical Keyring
> that uses the common Cache.

If the Cache Entry is still valid
(the TTL has not expired),

> it can be served without any KMS or DDB requests.

Reviewer 4: Do we have any concrete example? Anything that has to do with GC gets tricky and can be source of hard to debug bugs.

- Credentials used when creating the KMS Client, and thus used when calling KMS
- Other properties of the KMS Client, such as the region, or request headers

## Background

A [Cryptographic Materials Cache](https://github.com/awslabs/aws-encryption-sdk-specification/blob/master/framework/cryptographic-materials-cache.md) can be used by a Hierarchy Keyring, a Caching CMM, and the DB-ESDK.
It can store four types of [materials](https://github.com/aws/aws-cryptographic-material-providers-library/blob/c07a51fc29ff70411f7573bca96d2a091db8c1ed/AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/Model/cryptographic-materials-cache.smithy#L89) for three different use-cases:

- Hierarchy Keyring: keyStore#BranchKeyMaterials
- CachingCMM: EncryptionMaterials and DecryptionMaterials

> - DB-ESDK: keyStore#BeaconKeyMaterials

Reviewer: Style: `#` is an odd joiner in Markdown. Are you trying to create a link?

Author: No, fixed it. Thanks.

These materials have certain cache identifiers for accessing
them until a TTL. As of [today](#today), the cache identifiers
are set up such that:

- Only one Hierarchy Keyring can use one cache,
  i.e., currently, we do NOT support a shared cache
  between multiple Hierarchy Keyrings.
- Multiple CachingCMMs can use a single shared cache
  (this is for native languages, there is currently no
  CachingCMM implemented in dafny)

> - We do NOT support a shared cache between multiple
>   DB-ESDK clients.

Reviewer: Issue/Suggestion/Blocking: The DB-ESDK client can, theoretically, consume CachingCMMs and Hierarchy Keyrings.

Where you write DB-ESDK, can you replace it with Searchable Encryption?

```suggestion
- We do NOT support a shared cache between multiple
  Searchable Encryption Configurations.
```

Author: I did not know that. Thanks for pointing this out.

Author: Updated this in the doc wherever I mentioned DB-ESDK

"Easy to Use" in a multiple KMS Relationship
environment.

## Summary

Relevant sim [ticket](https://sim.amazon.com/issues/CrypTool-5311).

To improve the MPL Consumer data key caching experience,
when using the Hierarchy Keyring,
we need to allow caching across KeyStores/KMS Clients/KMS Keys.

To facilitate Caching across KeyStores/KMS Clients/KMS Keys,
we MUST break the Cryptographic Materials Cache (CMC)
out of the Hierarchy Keyring.

By allowing MPL Consumers to provide an already initialized CMC
to the Hierarchy Keyring at construction,
the CMC MAY cache Branch Keys protected by different
KMS Relationships.

This simplifies Multiple KMS Relationship MPL Consumers,

> as they do not need to stand up LRU Cache of Hierarchy Keyrings.

Reviewer: ```suggestion
as they do not need to stand up a LRU Cache of Hierarchy Keyrings.
```

Instead, they may maintain one CMC.
They still create a Hierarchy Keyring instance per KMS Relationship,
and they MUST use the correct Keyring to retrieve material
from the Cache.

> But they only need not maintain the common cache.

Reviewer: ```suggestion
But they only need to maintain the common cache.
```

Additionally, even though there is no CachingCMM in dafny right
now, in addition to allowing a shared cache between multiple
Hierarchy Keyrings, we want to create cache identifiers such
that multiple CachingCMMs and multiple Hierarchy Keyrings can

> simultaneously use a single shared cache.

Reviewer: Issue: This paragraph is hard to read.

Can you re-write it? 
I think the summary is:
> In the future, the CachingCMM will be introduced to Crypto Tool's Dafny products.
The CachingCMM and the Hierarchy Keyring both consume CMCs.
Thus, it will be possible to provide a CMC to both a Hierarchy Keyring & a CachingCMM.
This design MUST consider how the cache entries of multiple Hierarchy Keyrings and CachingCMMs will be appropriately separated.

## Requirements

- We must allow caching across KeyStores/KMS Clients/KMS Keys
  for multiple Hierarchy Keyrings and CachingCMMs.
  For this, we need to strategically update the cache identifiers
  for all materials ([background](#background)) in the CMC.
  The [Issues and Alternatives section](#issues-and-alternatives)
  discusses this.

## Out of Scope

> - Supporting a shared CMC for multiple KMS relationships
>   in the **DB-ESDK** to fetch BeaconKeyMaterials.

Reviewer: ```suggestion
- Cache Across multiple Searchable Encryption Configurations, 
  or Searchable Encryption's caching allowing for  multiple KMS relationships.
```

## Issues and Alternatives:

Before looking at the issues, please take a quick look at
[Hierarchy Keyring Cache Identifier formula](#hierarchy-keyring-cache-identifier-formula),
[CachingCMM Cache Identifier formula](#cachingcmm-cache-identifier-formula), and

> [DB-ESDK Cache Identifier formula](#db-esdk-cache-identifier-formula) for more context.

Reviewer: Copy these over.
Once this work lands, the content will be different.
If you need the current formula's for this change doc to make sense, 
you will need to statically define them.

Author: I copied the CachingCMM cache identifier formula. I've tried to reference the current commit wherever possible. Thanks for pointing this out.

If not, the recreated Hierarchy Keyring
is used to refresh it.

## A2: Sub-requirements

- We must update the cache identifier for [Hierarchy Keyring Branch Key Materials (Decryption) cache identifier](https://github.com/awslabs/aws-encryption-sdk-specification/blob/master/framework/aws-kms/aws-kms-hierarchical-keyring.md#decryption-materials) to have a branch-key-digest=SHA512(branch-key-id) appended, instead of a branch-key-id, just like in the [Branch Key Materials (Encryption) cache identifier](https://github.com/awslabs/aws-encryption-sdk-specification/blob/master/framework/aws-kms/aws-kms-hierarchical-keyring.md#encryption-materials) [[ref: Hierarchy Keyring Cache Identifier formula]](#hierarchy-keyring-cache-identifier-formula).

> - Minor update to the spec: The branch-key-version (which is the UUID) in the [Hierarchy Keyring Branch Key Materials (Decryption) cache identifier](https://github.com/awslabs/aws-encryption-sdk-specification/blob/master/framework/aws-kms/aws-kms-hierarchical-keyring.md#decryption-materials) is currently mentioned to be of size 36 bytes. However, the UUID is actually only 16 bytes of randomness. We get 36 bytes when we convert the 16 bytes of randomness in UUID to UTF8 encoding. Therefore, the spec should mention that the branch-key-version has size 16 bytes [[ref: Hierarchy Keyring Cache Identifier formula]](#hierarchy-keyring-cache-identifier-formula).

Reviewer: Suggestion: I know I used the term "entropy" and  "randomness", but technically, there is UUIDs are not a secure source of randomness.
 
```suggestion
- Minor update to the spec: The branch-key-version (which is the UUID) in the [Hierarchy Keyring Branch Key Materials (Decryption) cache identifier](https://github.com/awslabs/aws-encryption-sdk-specification/blob/master/framework/aws-kms/aws-kms-hierarchical-keyring.md#decryption-materials) is currently mentioned to be of size 36 bytes. However, the UUID is actually only 16 bytes. We get 36 bytes when we convert the 16 bytes in UUID to a UTF8 encoding. Therefore, the spec should mention that the branch-key-version has size 16 bytes [[ref: Hierarchy Keyring Cache Identifier formula]](#hierarchy-keyring-cache-identifier-formula).
```

Author: Okay, thanks for pointing this out!
